cava/CoqStyleGuide.md

```diff
@@ -0,0 +1,242 @@
+  ... | ]` before the `.` to explicitly specify how many subgoals you expect.
+  * Examples: `split; [ | ].` `induction z; [ | | ].`
+  * This helps make code more maintainable, because it fails immediately if
+    your tactic no longer solves as many subgoals as expected (or unexpectedly
+solves more).
+- If invoking a string of tactics (composed by `;`) that will break the goal into multiple subgoals and then solve all but one, still use `[ ]` to enforce that all but one goal is solved.
+  * Example: `split; try lia; [ ]`.
+- Tactics that consist only of `repeat`ing a procedure (e.g. `repeat match`,
+  `repeat first`) should factor out a single step of that procedure a separate
+tactic called `<tactic name>_step`, because the single-step version is much
+easier to debug. For instance:
+  ```coq
+  Ltac crush_step :=
+    match goal with
+    | _ => progress subst
+    | _ => reflexivity
+    end.
+  Ltac crush := repeat crush_step.
+  ```
+
+## Naming
+
+- Helper proofs about standard library datatypes should go in a module that is
+  named to match the standard library module (see example).
+  * This makes the helper proofs look like standard-library ones, which is
+    helpful for categorizing them if they're genuinely at the standard-library
+    level of abstraction.
+- Names of modules should start with capital letters.
+- Names of inductives and their constructors should start with capital letters.
+
+## Example
+
+A small standalone Coq file that exhibits many of the style points.
+
+```coq
+(****************************************************************************)
+(* Copyright 2020 The Project Oak Authors                                   *)
+(*                                                                          *)
+(* <License blurb..................................>                        *)
+(* <...............................................>                        *)
+(****************************************************************************)
+
+Require Import Coq.Lists.List.
+Require Import Coq.micromega.Lia.
+Require Import Coq.ZArith.ZArith.
+Import ListNotations.
+Local Open Scope Z_scope.
+
+(* Helper proofs about standard library integers (Z) go within [Module Z] so
+   that they match standard-library Z lemmas when used. *)
+Module Z.
+  Lemma pow_3_r x : x ^ 3 = x * x * x.
+  Proof. lia. Qed. (* very short proofs can go all on one line *)
+
+  Lemma pow_4_r x : x ^ 4 = x * x * x * x.
+  Proof.
+    change 4 with (Z.succ (Z.succ (Z.succ (Z.succ 0)))).
+    repeat match goal with
+           | _ => rewrite Z.pow_1_r
+           | _ => rewrite Z.pow_succ_r by lia
+           | |- context [x * (?a * ?b)] =>
+             replace (x * (a * b)) with (a * b * x) by lia
+           | _ => reflexivity
+           end.
+  Qed.
+End Z.
+(* Now we can access the lemmas above as Z.pow_3_r and Z.pow_4_r, as if they
+   were in the ZArith library! *)
+
+Definition bar (x y : Z) := x ^ (y + 1).
```

Comment:
Any style guidelines on when to give the result type explicitly? I see it is implicit here.

Reply:
I'm personally all over the place with when I give the result type explicitly, except that I generally do it less often for simple types and more often for complex types. But maybe we should come up with a norm for this. I wouldn't be opposed to declaring that we always write the type explicitly.

Reply:
I agree that it is best to just use judgement here and omit the type in simple cases, but provide it in other cases where the return type is not totally obvious.

Reply:
Sounds good. I'll add a bullet point somewhere for that.